srtcore/group.cpp

```diff
@@ -2996,89 +3100,33 @@ bool CUDTGroup::sendBackup_CheckRunningStability(const gli_t d, const time_point
-        const steady_clock::duration td_responsive = currtime - u.m_tsLastRspTime;
-        bool check_stability = true;
-
-        if (!is_zero(u.m_tsFreshActivation) && u.m_tsFreshActivation < currtime)
-        {
-            // The link is temporary-activated. Calculate then since the activation time.
-
-            // Check the last received ACK time first. This time is initialized with 'now'
-            // at the CUDT::open call, so you can't count on the trap zero time here, but
-            // it's still possible to check if activation time predates the ACK time. Things
-            // are here in the following possible order:
-            //
-            // - ACK time (old because defined at open)
-            // - Response time (old because the time of received handshake or keepalive counts)
-            // ... long time nothing ...
-            // - Activation time.
-            //
-            // If we have this situation, we have to wait for at least one ACK that is
-            // newer than activation time. However, if in this situation we have a fresh
-            // response, that is:
-            //
-            // - ACK time
-            // ...
-            // - Activation time
-            // - Response time (because a Keepalive had a caprice to come accidentally after sending)
-            //
-            // We still wait for a situation that there's at least one ACK that is newer than activation.
-
-            // As we DO have activation time, we need to check if there's at least
-            // one ACK newer than activation, that is, td_acked < td_active
-            if (u.m_tsLastRspAckTime < u.m_tsFreshActivation)
-            {
-                check_stability = false;
-                HLOGC(gslog.Debug,
-                      log << "grp/sendBackup: link @" << d->id
-                          << " activated after ACK, "
-                             "not checking for stability");
-            }
-            else
-            {
-                u.m_tsFreshActivation = steady_clock::time_point();
-            }
-        }
-
-        if (check_stability && count_microseconds(td_responsive) > m_uOPT_StabilityTimeout)
-        {
-            if (is_zero(u.m_tsUnstableSince))
-            {
-                HLOGC(gslog.Debug,
-                      log << "grp/sendBackup: socket NEW UNSTABLE: @" << d->id << " last heard "
-                          << FormatDuration(td_responsive) << " > " << m_uOPT_StabilityTimeout
-                          << " (stability timeout)");
-                // The link seems to have missed two ACKs already.
-                // Qualify this link as unstable
-                // Notify that it has been seen so since now
-                u.m_tsUnstableSince = currtime;
-            }
-
-            is_stable = false;
-        }
-    }
+    const int is_stable = sendBackup_CheckRunningLinkStable(u, currtime, d->weight);
```

Comment:
How do you protect yourself against rapid changes in this value in case when, for example, RTT starts to increase dramatically?

Reply:
RTT changes are smoothed by the IIR filter (first on the receiver side, then again on the sender side). So there won't be that high fluctuation. And if some fluctuations fall outside the stability threshold, that is an indicator that something unstable is happening on the link.